File: 005 Longest Palindromic Substring/longest_palindromic_substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        if not s or len(s) <= 1: # base case
            return s
        start = end = 0
        for i in (range(len(s))):
            len1 = self.expand(s, i, i) # center is i
            len2 = self.expand(s, i, i + 1) # center is between i and i+1
            length = max(len1, len2)
            if (length > end - start): # longer substring
                start = i - (length - 1) // 2
                end = i + length // 2
        return s[start:end + 1]
    
    def expand(self, s, l, r):
        while (l >= 0 and r < len(s) and s[l] == s[r]): # expand from center
            l -= 1
            r += 1
        return r - l - 1

File: 5 Longest Palindromic Substring/LongestPalindromicSubstring.java
class Solution {
    public String longestPalindrome(String s) {
        if (s == null || s.length() <= 1) {
            return s;
        }
        int n = s.length();
        int start = 0;
        int end = 0;
        for (int i = 0; i < n; i++) {
            int len1 = expand(s, i, i); // center is i
            int len2 = expand(s, i, i + 1); // center is between i and i+1
            int len = Math.max(len1, len2);
            if (len > end - start) {    // find the place of the substring
                start  = i - (len - 1) / 2; 
                end = i + len / 2;
            }
        }
        return s.substring(start, end + 1);
    }
    
    private int expand(String s, int left, int right) {
        int l = left;
        int r = right;
        int n = s.length();
        while (l >= 0 && r < n && s.charAt(l) == s.charAt(r)) { // expand
            l--;
            r++;
        }
        return r - l - 1;
    }
}